### PR TITLE
[rpc] don't crash callee when function does not exist on it, instead return Exception

### LINKED ISSUE #27368: Improve error message if Python function is not found on callee instead of crashing
@rohan-varma found that when the target Python function is missing on the callee side when using `rpc`/`remote`, the callee will hit segfault instead of sending the error message back to the caller. 

```python
# On caller
import torch
import torch.distributed as dist

dist.init_process_group("gloo", init_method='tcp://127.0.0.1:12356', world_size=2, rank=1)
dist.init_model_parallel("worker0")

def my_tensor_function(a, b):
    return a + b

ret = dist.rpc_sync("worker1", my_tensor_function, args=(torch.ones(2, 2), torch.ones(2, 2)))
dist.join_rpc()
```

```python
# On callee
import torch.distributed as dist

#def my_tensor_function(a, b):
#        return a + b

dist.init_process_group("gloo", init_method='tcp://127.0.0.1:12356', world_size=2, rank=0)
dist.init_model_parallel("worker1")
dist.join_rpc()
```

For example, the above program will show the following error messages on the callee:

```
Thread 6 "python" received signal SIGSEGV, Segmentation fault.
[Switching to LWP 4177163]
PyErr_Fetch (p_type=0x7fffb37fd740, p_value=0x7fffb37fd748, p_traceback=0x7fffb37fd750) at /tmp/build/80754af9/python_1553721932202/work/Python/errors.c:339          
339     /tmp/build/80754af9/python_1553721932202/work/Python/errors.c: No such file or directory.                                                                     
(gdb) backtrace
#0  PyErr_Fetch (p_type=0x7fffb37fd740, p_value=0x7fffb37fd748, p_traceback=0x7fffb37fd750) at /tmp/build/80754af9/python_1553721932202/work/Python/errors.c:339      
#1  0x00007fffe26bca53 in pybind11::error_already_set::~error_already_set() () from /data/users/shenli/pytorch/torch/lib/libtorch_python.so                           
#2  0x00007fffe23ce0bc in __gxx_exception_cleanup (code=<optimized out>, exc=0x7fffa0000e40)                                                                          
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/libsupc++/eh_throw.cc:49                
#3  __gxx_exception_cleanup (code=<optimized out>, exc=<optimized out>)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/libsupc++/eh_throw.cc:33                
#4  0x00007fffd2d2314b in torch::distributed::rpc::RequestCallback::operator()(torch::distributed::rpc::Message&) const ()                                            
   from /data/users/shenli/pytorch/torch/lib/libtorch.so
#5  0x00007fffe2be5556 in torch::distributed::rpc::ProcessGroupAgent::enqueueRecv(torch::distributed::rpc::RecvWork)::{lambda(torch::distributed::rpc::RecvWork&)#1}::operator()(torch::distributed::rpc::RecvWork&) const () from /data/users/shenli/pytorch/torch/lib/libtorch_python.so                                                   
#6  0x00007fffcc0cff13 in c10::ThreadPool::main_loop(unsigned long) () from /data/users/shenli/pytorch/torch/lib/libc10.so                                            
#7  0x00007fffe23ea421 in std::execute_native_thread_routine_compat (__p=<optimized out>)                                                                             
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/src/c++11/thread.cc:94                  
#8  0x00007ffff7bc6dd5 in start_thread () from /lib64/libpthread.so.0
#9  0x00007ffff78efead in clone () from /lib64/libc.so.6
```

cc @ezyang @gchanan @zou3519 @jerryzh168 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

### PR DESCRIPTION
**Pull request by rohan-varma**
Closes https://github.com/pytorch/pytorch/issues/27368. 
Previously, if a function `'func` did not exist on worker A but existed in B, and the user ran `rpc.rpc_sync(A,  func)`, A would crash with a segmentation fault since it is not able to find the function. B would eventually timeout since RPCs by default time out in 60s.

At the root this comes from an unhandled exception when trying to deserialize the `PythonUDF` to run.

This PR makes it so that we can recover from this error, and A reports back a `RemoteException` to B indicating that the function was not found. Now, A will no longer crash and B can handle the exception appropriately and with more information.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit bf321176:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=bf3211762a7dd0cd11f010f4008d6853d84b0633).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4362814) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92191940) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92191940)) :snowflake:
<details>
<summary>
<code>Jan 28 22:32:23 RuntimeError: quantization scale should be > 0</code>
</summary>

```
Jan 28 22:32:23   File "test_quantization.py", line 730, in test_quantized_rnn 
Jan 28 22:32:23     y, (h, c) = cell_dq(x, (h, c)) 
Jan 28 22:32:23   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 540, in __call__ 
Jan 28 22:32:23     result = self.forward(*input, **kwargs) 
Jan 28 22:32:23   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 373, in forward 
Jan 28 22:32:23     return self.forward_tensor(input, hx) 
Jan 28 22:32:23   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 334, in forward_tensor 
Jan 28 22:32:23     input, hx, batch_sizes, max_batch_size, sorted_indices) 
Jan 28 22:32:23   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 315, in forward_impl 
Jan 28 22:32:23     self.batch_first, dtype=self.dtype, use_dynamic=True) 
Jan 28 22:32:23 RuntimeError: quantization scale should be > 0 
Jan 28 22:32:23  
Jan 28 22:32:23 ---------------------------------------------------------------------- 
Jan 28 22:32:23 Ran 36 tests in 54.907s 
Jan 28 22:32:23  
Jan 28 22:32:23 FAILED (errors=1, skipped=1) 
Jan 28 22:32:23  
Jan 28 22:32:23 Generating XML reports... 
Jan 28 22:32:23 Traceback (most recent call last): 
Jan 28 22:32:23   File "test/run_test.py", line 456, in <module> 
Jan 28 22:32:23     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 1]
```diff
@@ -1626,3 +1630,19 @@ class TestPickler():
...
+
+    @dist_init
+    def test_function_not_on_callee(self):
+        # test that if a function does not exist on a callee, we don't crash,
+        # instead we get an AttributeError indicating that the func does not exist.
+        if self.rank != 0:
+            # Use delattr to remove the binding of a func on callee nodes
+            import sys
+            this_module = sys.modules[__name__]
+            delattr(this_module, "foo_add")
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: Question: I wonder do we have to explicitly delete the attr? Does it work if we define `foo_add` as a local function within the test?

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks!

**Comment by mrshenli** [comment]
There is a conflict in `torch/testing/_internal/distributed/rpc/rpc_test.py`, need to address that before landing.

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 1 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: @mrshenli If we define the function locally, then on the caller side (worker0 in this case) the pickler fails with a `Cannot pickle local function` error. So it looks like that case already has error handling.

**Pull request closed without merging**